### PR TITLE
Get more information from windows events

### PR DESCRIPTION
**Pull request by achntrl**
### What does this PR do?

Windows events will now display a full text message instead of a JSON
object. When available, the agent will now enrich the events with status,
human readable task name and opcode. 

### Motivation

We want to be able display human readable information in the message field of Log Explorer instead of a JSON object

### Additional Notes

Used those docs:

- https://docs.microsoft.com/en-us/windows/win32/wes/rendering-events
- https://docs.microsoft.com/en-us/windows/win32/wes/formatting-event-messages

### CODE REVIEW

File: pkg/logs/input/windowsevent/tailer.go [hunk 1]
```diff
@@ -70,7 +79,8 @@ func (t *Tailer) Identifier() string {
 }
 
 // toMessage converts an XML message into json
-func (t *Tailer) toMessage(event string) (*message.Message, error) {
+func (t *Tailer) toMessage(rm *richEvent) (*message.Message, error) {
```
Comment (blemale) [inline pkg/logs/input/windowsevent/tailer.go]:
```suggestion
func (t *Tailer) toMessage(re *richEvent) (*message.Message, error) {
```

**Review by blemale — COMMENTED** [review]
For my knowledge, do you have some documentation on the Windows event logging thing?

File: pkg/logs/input/windowsevent/event.h [hunk 2]
```diff
@@ -3,14 +3,25 @@
...
 
+typedef struct RichEvent_t {
+    LPWSTR message;
+    LPWSTR task;
+    LPWSTR opcode;
+    LPWSTR level;
+} RichEvent ;
+
 ULONGLONG startEventSubscribe(char *channel, char* query, ULONGLONG  hBookmark, int flags, PVOID ctx);
+RichEvent* EnrichEvent(ULONGLONG ulEvent);
```
[the remark this answers is not in the document]
Comment (achntrl) [reply pkg/logs/input/windowsevent/event.h]: Yes

File: pkg/logs/input/windowsevent/eventlog.c [hunk 3]
```diff
@@ -147,3 +147,165 @@ DWORD PrintEvent(EVT_HANDLE hEvent)
...
 	return status;
 }
+
+LPWSTR FormatEvtField(EVT_HANDLE hMetadata, EVT_HANDLE hEvent, EVT_FORMAT_MESSAGE_FLAGS FormatId);
+PEVT_VARIANT GetProviderName(EVT_HANDLE hEvent);
+RichEvent* EnrichEvent(ULONGLONG ulEvent)
+{
+    EVT_HANDLE hProviderMetadata = NULL;
+    LPWSTR pwsMessage = NULL;
+    EVT_HANDLE hEvent = (EVT_HANDLE) ulEvent;
```
[the remark this answers is not in the document]
Comment (achntrl) [reply pkg/logs/input/windowsevent/eventlog.c]: EVT_HANDLE is a type def for a HANDLE which is a pointer (void *)

iirc, to keep compatibility with go type we cast it to a ULONGLONG

File: pkg/logs/input/windowsevent/tailer_windows.go [hunk 4]
```diff
@@ -183,3 +229,22 @@ func ConvertWindowsString(winput []uint8) string {
...
+
+// LPWSTRToString converts a C.LPWSTR to a string. It also truncates the
+// strings to 128kB as a basic protection mechanism to avoid allocating an
+// array too big. Messages with more than 128kB are likely to be bigger
+// than 256kB when serialized and then dropped
+func LPWSTRToString(cwstr C.LPWSTR) string {
+	ptr := unsafe.Pointer(cwstr)
+	sz := C.wcslen((*C.wchar_t)(ptr))
+	sz = min(sz, maxRunes)
+	wstr := (*[maxRunes]uint16)(ptr)[:sz:sz]
```
[the remark this answers is not in the document]
Comment (achntrl) [reply pkg/logs/input/windowsevent/tailer_windows.go]: Done here https://github.com/DataDog/datadog-agent/pull/3987/commits/377eee8690440cd9471a2bab6301ddbdfdbd6775

**Comment by achntrl** [comment]
> For my knowledge, do you have some documentation on the Windows event logging thing?

@blemale I've put 2 links in the description about the rendering / formatting

The method used for subscription is document here: https://docs.microsoft.com/en-us/windows/win32/api/winevt/nf-winevt-evtsubscribe

It takes a callback that we define in go and export to C `goNotificationCallback` that calls `EvtRender` that now calls the new `enrichEvent` method

**Review by derekwbrown — DISMISSED** [review]
Looks good to me.  I would have one stylistic comment (which you may choose to ignore).
I would avoid having 1-line `if` statements without `{}`.  So
```
if (a)
    b
```
while syntactically correct, should be
```
if (a) {
  b
}
```

**Review by derekwbrown — APPROVED** [review]

**Review by gaetan-deputier — APPROVED** [review]

**Pull request merged**